stan/math/rev/functor/cvodes_integrator_adjoint.hpp

```diff
@@ -0,0 +1,829 @@
+   * @param[in] t Time
+   * @param[out] J CVode structure where output is to be stored
+   */
+  inline void jacobian_adj(double t, N_Vector y, SUNMatrix J) const {
+    Eigen::Map<Eigen::MatrixXd> J_adj_y(SM_DATA_D(J), N_, N_);
+
+    // J_adj_y = -1 * transpose(J_y)
+    jacobian_states(t, y, J);
+
+    J_adj_y.transposeInPlace();
+    J_adj_y.array() *= -1.0;
+  }
+
+  /**
+   * Overloads which setup the states returned from the forward solve. In case
+   * the return type is a double only, then no autodiff is needed. In case of
+   * autodiff then non-chaining varis are setup accordingly.
+   */
+  void store_state(std::size_t n, const Eigen::VectorXd& state,
+                   Eigen::Matrix<var, Eigen::Dynamic, 1>& state_return) {
+    solver_->y_[n] = state;
+    state_return.resize(N_);
+    for (size_t i = 0; i < N_; i++) {
+      non_chaining_varis_[N_ * n + i] = new vari(state.coeff(i), false);
+      state_return.coeffRef(i) = var(non_chaining_varis_[N_ * n + i]);
+    }
+  }
+
+  void store_state(std::size_t n, const Eigen::VectorXd& state,
+                   Eigen::Matrix<double, Eigen::Dynamic, 1>& state_return) {
+    solver_->y_[n] = state;
+    state_return = state;
+  }
+
+  /**
+   * Since the CVODES solver manages memory with malloc calls, these resources
+   * must be freed using a destructor call (which is not being called for the
+   * vari class).
+   */
+  struct cvodes_solver : public chainable_alloc {
+    const size_t N_;
+    std::vector<Eigen::VectorXd> y_;
+    const std::string function_name_str_;
+    const char* function_name_;
+    void* cvodes_mem_;
+    N_Vector nv_state_forward_;
+    N_Vector nv_state_backward_;
+    N_Vector nv_quad_;
+    N_Vector nv_absolute_tolerance_forward_;
+    N_Vector nv_absolute_tolerance_backward_;
+
+    SUNMatrix A_forward_;
+    SUNLinearSolver LS_forward_;
+
+    SUNMatrix A_backward_;
+    SUNLinearSolver LS_backward_;
+
+    template <typename StateFwd, typename StateBwd, typename Quad,
+              typename AbsTolFwd, typename AbsTolBwd>
+    cvodes_solver(const char* function_name, size_t N, size_t num_args_vars,
+                  size_t ts_size, int solver_forward, StateFwd& state_forward,
+                  StateBwd& state_backward, Quad& quad,
+                  AbsTolFwd& absolute_tolerance_forward,
+                  AbsTolBwd& absolute_tolerance_backward)
+        : chainable_alloc(),
+          N_(N),
+          y_(ts_size),
+          function_name_str_(function_name),
+          function_name_(function_name_str_.c_str()) {
+      nv_state_forward_ = N_VMake_Serial(N, state_forward.data());
+      nv_state_backward_ = N_VMake_Serial(N, state_backward.data());
+      nv_quad_ = N_VMake_Serial(num_args_vars, quad.data());
+      nv_absolute_tolerance_forward_
+          = N_VMake_Serial(N, absolute_tolerance_forward.data());
+      nv_absolute_tolerance_backward_
+          = N_VMake_Serial(N, absolute_tolerance_backward.data());
```

Comment:
Should move these to the constructor initializer. I think for `LS_forward_` is there default initialization just to be a nullptr?

Reply:
ok